File: Bulgarian Competitions/BNOI - Group D/Points/Points.cs
using System;
using System.Linq;

class RightWord
{
    static void Main()
    {
        int N = int.Parse(Console.ReadLine());
        int[,] matrix = new int[N, N];
        for (int i = 0; i < N; i++)
        {
            string[] line = Console.ReadLine().Split(' ');
            for (int j = 0; j < N; j++)
            {
                matrix[i, j] = int.Parse(line[j]);
            }
        }

        int points = 0;
        int[] primeDiagonal = new int[N];
        int[] secondDiagonal = new int[N];
        for (int i = 0; i < N; i++)
        {
            int[] rows = new int[N];
            int[] cols = new int[N];
            for (int j = 0; j < N; j++)
            {
                rows[j] = matrix[i, j];
                cols[j] = matrix[j, i];
                if (i == j) primeDiagonal[i] = matrix[i, j];
                if (i + j == N - 1) secondDiagonal[i] = matrix[i, j];
            }
            points += FindSequence(rows, "row");
            points += FindSequence(cols, "col");
        }

        points += FindSequence(primeDiagonal, "diagonal1");
        points += FindSequence(secondDiagonal, "diagonal2");
        Console.WriteLine(points);
    }

    static int FindSequence(int[] array, string type)
    {
        var unique = array.Distinct().ToArray();
        Array.Sort(unique);
        int result = 0;
        int count = 1;
        for (int i = 1; i < unique.Length; i++)
        {
            if (unique[i] == unique[i - 1] + 1) count++;
            else
            {
                result += Points(count, type);
                count = 1;
            }
        }

        return result + Points(count, type);
    }

    static int Points(int count, string type)
    {
        switch (count)
        {
            case 1: return 0;
            case 2: return (type == "row") ? 10 : ((type == "col") ? 20 : 30);
            case 3: return (type == "row") ? 40 : ((type == "col") ? 50 : 60);
            case 4: return (type == "row") ? 70 : ((type == "col") ? 80 : 90);
            case 5: return (type == "row") ? 100 : ((type == "col") ? 110 : 120);
            default: return (type == "row") ? 150 : ((type == "col") ? 160 : 170);
        }
    }
}

File: Bulgarian Competitions/BNOI - Group D/Points/Points.js
function solve(params) {
    var args = params[0].split('\n');
    var N = parseInt(args[0]);
    var matrix = [];
    for (var i = 0; i < N; i++) {
        matrix.push(args[i + 1].split(' ').map(Number));
    }

    var points = 0;
    var primeDiagonal = [];
    var secondDiagonal = [];
    for (var i = 0; i < N; i++) {
        var rows = [];
        var cols = [];
        for (var j = 0; j < N; j++) {
            rows.push(matrix[i][j]);
            cols.push(matrix[j][i]);
            if (i == j) primeDiagonal.push(matrix[i][j]);
            if (i + j == N - 1) secondDiagonal.push(matrix[i][j]);
        }
        points += FindSequence(rows, "row");
        points += FindSequence(cols, "col");
    }

    points += FindSequence(primeDiagonal, "diagonal1");
    points += FindSequence(secondDiagonal, "diagonal2");
    return points;

    function FindSequence(array, type) {
        Array.prototype.distinct = function () {
            var a = [], k = 0, e;
            for (k = 0; e = this[k]; k++)
                if (a.indexOf(e) == -1)
                    a.push(e);
            return a;
        }

        var unique = array.distinct().sort(function (a, b) { return a - b });
        var result = 0;
        var count = 1;
        for (var i = 1; i < unique.length; i++) {
            if (unique[i] == unique[i - 1] + 1) count++;
            else {
                result += Points(count, type);
                count = 1;
            }
        }

        return result + Points(count, type);
    }

    function Points(count, type) {
        switch (count) {
            case 1: return 0;
            case 2: return (type == "row") ? 10 : ((type == "col") ? 20 : 30);
            case 3: return (type == "row") ? 40 : ((type == "col") ? 50 : 60);
            case 4: return (type == "row") ? 70 : ((type == "col") ? 80 : 90);
            case 5: return (type == "row") ? 100 : ((type == "col") ? 110 : 120);
            default: return (type == "row") ? 150 : ((type == "col") ? 160 : 170);
        }
    }
}